### ISSUE TITLE
Thread getting blocked while destroying atomic long instances

### ISSUE THREAD

**Issue by harsha89** [hazelcast/hazelcast#7078]
Hi,

We have a cluster which expands dynamically. We are using atomic long to maintain the global counters. Since atomic longs doesn't have TTL, we are writing a cleanup task to clean expired instances. But when task is running in mulltiple instances. It's tend to hang. Giving following thread dump. 

"Throttle Distributed Instance Cleanup Task" daemon prio=10 tid=0x00007fda3446a000 nid=0x12a4 in Object.wait() [0x00007fda294c4000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000007d5bc6550> (a com.hazelcast.spi.impl.operationservice.impl.InvocationFuture)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.pollResponse(InvocationFuture.java:299)
	- locked <0x00000007d5bc6550> (a com.hazelcast.spi.impl.operationservice.impl.InvocationFuture)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.waitForResponse(InvocationFuture.java:247)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.get(InvocationFuture.java:224)
	at com.hazelcast.util.FutureUtil.executeWithDeadline(FutureUtil.java:294)
	at com.hazelcast.util.FutureUtil.waitWithDeadline(FutureUtil.java:278)
	at com.hazelcast.util.FutureUtil.waitWithDeadline(FutureUtil.java:252)
	at com.hazelcast.spi.impl.proxyservice.impl.ProxyServiceImpl.destroyDistributedObject(ProxyServiceImpl.java:143)
	at com.hazelcast.spi.AbstractDistributedObject.destroy(AbstractDistributedObject.java:55)
	at org.wso2.carbon.throttle.core.SharedParamManager.removeTimestampWithFullId(SharedParamManager.java:252)
	at org.wso2.carbon.throttle.core.ThrottleDistributedInstancesCleanupTask$CleanupTask.run(ThrottleDistributedInstancesCleanupTask.java:128)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:441)
	at java.util.concurrent.FutureTask$Sync.innerRunAndReset(FutureTask.java:317)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:150)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$101(ScheduledThreadPoolExecutor.java:98)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.runPeriodic(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:204)
	at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)
	at java.lang.Thread.run(Thread.java:662)

Thread related to : https://groups.google.com/forum/#!topic/hazelcast/ILKCEZX14b8

Thanks,
Harsha

**Comment by jerrinot** [comment]
Hello @harsha89,

thanks for reporting the issue. What version of Hazelcast do you use? Could you share source code of your clean-up tasks and way how you are submitting them?

Thanks!

**Comment by harsha89** [comment]
The hazelcast version is 3.5.0.
During the cleanup task, each node first take all atomic longs for particular service, the execute the .destroy method. Thread does not get hang when cleanup  is running within a single machine. But if every node in cluster run the cleanup task, then we can observe thread getting hanged.

**Comment by jerrinot** [comment]
@harsha89: Thanks for your update. Could you share your code to submit the clean-up tasks? Is the task targeted to a member owning the counter?

**Comment by harsha89** [comment]
No it's not owning to targeted member. The issues is that, we need to cleanup the atomic longs because it's not have expiry. Here the overall code
Collection<DistributedObject> distributedObjects = hazelcastInstance.getDistributedObjects();

for (DistributedObject distributedObject : distributedObjects) {
                        serviceName = distributedObject.getServiceName();
                        if (AtomicLongService.SERVICE_NAME.equals(serviceName)) {
                            name = distributedObject.getName();
                            if (name.contains(ThrottleConstants.THROTTLE_TIMESTAMP_KEY)) {
                               		//SOme logic to find expired instances
                                    removeTimestampWithFullId(name);//calling .destroy in the instance
                                }
                            }
                        }
                    }

Thanks,
harsha

**Comment by metanet** [comment]
Hi @harsha89 

Can you please provide a full code to reproduce the issue? I tried to reproduce it on both master and 3.5 but it seems working.

Additionally, to destroy your atomic instances, it is sufficient to call ``destroy`` method on only a single hz instance. 

Based on my understanding of your report, here is the code I have tried:

```
import com.hazelcast.concurrent.atomiclong.AtomicLongService;
import com.hazelcast.core.DistributedObject;
import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.test.HazelcastTestSupport;
import com.hazelcast.test.TestHazelcastInstanceFactory;
import org.junit.Before;
import org.junit.Test;

import java.util.Collection;
import java.util.concurrent.CountDownLatch;

public class AtomicLongDestroyTest
        extends HazelcastTestSupport {

    private HazelcastInstance[] instances;

    @Before
    public void before() {
        final TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(4);
        instances = factory.newInstances();
    }

    @Test
    public void testAtomicLongDestroy() {
        for (int i = 0; i < 50000; i++) {
            instances[0].getAtomicLong(randomMapName("pref1"));
            instances[1].getAtomicLong(randomMapName("pref2"));
            instances[2].getAtomicLong(randomMapName("pref3"));
            instances[3].getAtomicLong(randomMapName("pref4"));
        }

        final CountDownLatch latch = new CountDownLatch(4);

        for (final HazelcastInstance instance : instances) {
            new Thread(new Runnable() {
                @Override
                public void run() {
                    final Collection<DistributedObject> objects = instance.getDistributedObjects();
                    for (DistributedObject obj : objects) {
                        if (obj.getServiceName().equals(AtomicLongService.SERVICE_NAME)) {
                            if (obj.getName().startsWith("pref1")) {
                                obj.destroy();
                            }
                        }
                    }
                    System.out.println("Counting down");
                    latch.countDown();
                }
            }).start();
        }

        assertOpenEventually(latch);
    }

}
```

**Comment by harsha89** [comment]
@metanet thanks for your insight on this. Is this test runs in parallel? We have several nodes which will  run cleanup task in random intervals and calls destroy. Each instance may not aware whether othernode called destroy concurrently.

**Comment by metanet** [comment]
Yes, the code above runs a separate task for each hazelcast instance and it works. It also calls ``destroy()`` on the same atomic long object from multiple instances. It should be same with your use case.
